mason init

ankingcodes opened the pull request:
Implementation of `mason init`, which creates a mason package in an existing directory with any partial package contents or metadata. A user can run with no arguments to initialize a package in the current directory, or provide a path to choose where to initialize a package.

Existing files are not overwritten by `mason init`. Existing files that pose a conflict with a mason package (such as a bad `Mason.toml`) will generate helpful error messages.

Includes testing and documentation updates

fixes #14786 
This is a feature also requested in #7106

ankingcodes commented:
@ben-albrecht @Spartee  please review this and give suggestions about what more should be taken care of. 👍 🚀

ankingcodes commented:
@ben-albrecht thanks for the suggestions !
>>I suspect many functions defined in MasonInit.chpl are defined elsewhere in mason source. It would be better to re-use those implementations

Yes, the `makeTomlFileFromName` exists as `makeBasicToml` in `Mason new` but it didn't work the way I wanted it to & I didn't want to edit that one because that might not be suitable for the `mason new` command. For example, the `makeBasicToml` function requires a `name` parameter which is the project-directory which is created by the new command. But the problem in `makeBasicToml` function is, 
```chapel
var tomlFile = open(name+"/Mason.toml", iomode.cw);
```
Here, its necessary that we are outside the project-directory. But in our `init command` we are already inside the project-directory.
Due to this, I have used a custom `makeTomlFileFromName`  function in `mason init`. 
>>See testing doc for getting started there.

Thanks for this one ! 

### Extras
-  Currently, the `mason init` creates a TOML file with the project-name as that of the directory it was used in / the name of the directory in the path provided along with the command. If a TOML file exists in a directory, it doesn't create another TOML file. Are there any improvements required in this?
- Should directories such as  `example/`, `src/`,`test/` be created ?
-  If directories are to be created, we might need a mechanism to check if file or directory already exists to avoid overwriting.

ben-albrecht commented:
> Yes, the `makeTomlFileFromName` exists as `makeBasicToml` in `Mason new` but it didn't work the way I wanted it to & I didn't want to edit that one because that might not be suitable for the 
> `mason new` command. For example, the `makeBasicToml` function requires a `name` parameter which is the project-directory which is created by the new command. But the problem in `makeBasicToml` function is, 
> ```chapel
> var tomlFile = open(name+"/Mason.toml", iomode.cw);
> ```
> Here, its necessary that we are outside the project-directory. But in our `init command` we are already inside the project-directory.

I see. I would be fine modifying the `MasonNew` version for the purposes of `MasonInit`. If you think about it, `mason init` can potentially end up doing everything `mason new` does (say if you initialized a mason project in an empty directory), so it would be a shame to re-implement all of `MasonNew` separately. We should instead aim to re-use as much of `MasonNew` as possible.

I would suggest the pattern of:
- Check if a thing exists (e.g. `Mason.toml`, `src/<file>.chpl`, `.git`, etc.)
  - If it does not, call the appropriate `MasonNew.chpl`-defined function to create it.
  - Otherwise, possibly validate it in some way and continue

For the case of [makeBasicToml](https://github.com/chapel-lang/chapel/blob/master/tools/mason/MasonNew.chpl#L116-L128), we can do one of the following:
- change our working directory up to 1 directory in `MasonInit` before calling it
- change our working directory into the package in `MasonNew` before calling it, and change the function to assume we are _in_ the package
- change the function to support a package `path=''` argument: if unset, assume we are in the package, if set, assume that path contains the package (so `MasonNew` calls would become `makeBasicToml(name, path=name)`

ben-albrecht commented:
> Currently, the mason init creates a TOML file with the project-name as that of the directory it was used in / the name of the directory in the path provided along with the command. If a TOML file exists in a directory, it doesn't create another TOML file. Are there any improvements required in this?

That's a good start. We should also make sure we can parse the existing TOML file and find a project name and version (at the very least).

> Should directories such as example/, src/,test/ be created ?

Yes, if they do not exist. I would expect the end result of a `mason init` to be similar to that of a `mason new`, barring anything unique the user has added.

> If directories are to be created, we might need a mechanism to check if file or directory already exists to avoid overwriting.

I agree. This will be a common pattern so we should definitely create a function to do this more concisely throughout the program.

ankingcodes commented:
@ben-albrecht 
> change the function to support a package path='' argument: if unset, assume we are in the package, if set, assume that path contains the package (so MasonNew calls would become makeBasicToml(name, path=name)

I have implemented this and it now works for both `mason new` and `mason init`.

> That's a good start. We should also make sure we can parse the existing TOML file and find a project name and version (at the very least).

This is unclear to me. How will the project name and version be used ?

ben-albrecht commented:
> I have implemented this and it now works for both mason new and mason init.

Nice! That turned out to be a much cleaner change than I expected.

> This is unclear to me. How will the project name and version be used ?

Here are some assertions to validate:

- Does the project name field exist in the TOML?
  - Does it have a valid value, i.e. it follows module identifier rules?
    - see `MasonNew.chpl:55-65` for existing checks -- we should encapsulate these checks into a re-usable function
   - If a `src/<module>.chpl` exists, does the project name match the `<file>` name?
- Does the version field exist in the TOML?
  - Does it have a valid value, e.g. `x.y.z`? (see `MasonModify.chpl:checkVersion()`

If a TOML file exists and fails any of these checks, I think we should report an error and halt.

ben-albrecht commented on tools/mason/MasonInit.chpl:
```diff
@@ -0,0 +1,76 @@
+				in the given path
+			*/
+			var path = name;
+			if(isDir(path)) {
```
I think this is reversed - we should throw an error if that directory does not already exist. The error could say something like:

```
"Directory does not exist: %s".format(path)
"Did you mean 'mason new' to create a new project from scratch?"
```

ankingcodes replied:
Shouldn't mason check if a target directory has been already created and if not created, it should create one. Dlang's DUB does this.

ben-albrecht replied:
I believe this is the main difference of `mason init` vs. `mason new`:

- `mason init` creates a mason project in an existing directory.
- `mason new` creates a new mason project from scratch.

If a user tried `mason init <path>` on a _path_ that does not exist, they should be using `mason new` instead.

ankingcodes replied:
Yeah. I understood that. That's why I deleted the previous comment 😜 
I'm working on it now. 👍

ben-albrecht commented on tools/mason/MasonInit.chpl:
```diff
@@ -0,0 +1,76 @@
+			var path = name;
+			if(isDir(path)) {
+					throw new owned MasonError("Path already exists.");
+			} else if (isFile(path + "/Mason.toml")) {
```
Similarly, it's OK if TOML file exists already, we just need to validate it, and throw an error if something is wrong with it.

ben-albrecht replied:
same for the `if name == ''` code path above

ankingcodes commented:
@ben-albrecht Thanks for the comments. I made the changes you requested. 
Currently, `mason init` works as follows : 

- `mason init` in the directory 
  - looks for files and directories already present and creates new files and directories without overwriting the previous ones.
  - if a `Mason.toml` file is found it parses projectName and version and validates it. 
- `mason init <target>` 
  -  looks for files and directories already present and creates new files and directories without overwriting the previous ones.
  - if a `Mason.toml` file is found it validates the projectName and version
 
I took care to reuse the functions already defined in Mason's source. There's a bug though. When we validate the TOML file for projectName, it uses a proc `validateAndInit` in `MasonNew.chpl (67)`. Here in line `79` , there's a check for `isDir(name)`. This creates a error for `mason init <target>` because a directory, obviously will exist. Please suggest what can be done about this. Should I create a separate function for this case ? We can however, use a hack and modify the `projectName` in `MasonInit.chpl (93)` to avoid this case, but that won't be very trivial right ?

ankingcodes commented:
`mason init` is almost done. Everything works fine except for the following :
consider the directory structure
```
project
      |--src
      |--test
      |--example
      |-- project2
            |-src
            |-test
            |--example
            |--project3 (empty)
```

if we perform `cd project/project2` and `mason init project3` --> works perfectly
however if we perform `cd project` and `mason init project2/project3` --> gives an error as follows.
```
uncaught MasonError: A directory named 'project3' already exists 
  ./MasonNew.chpl:80: thrown here
  ./MasonInit.chpl:93: uncaught here
```

Since we want to reuse functions of `MasonNew` , I used that. But this case isn't covered.
Also we cannot skip the condition `isDir(name)` because its a crucial one. How should I tackle this ?

ben-albrecht commented:
Thanks for your patience @ankingcodes --

> Since we want to reuse functions of MasonNew , I used that. But this case isn't covered.
Also we cannot skip the condition isDir(name) because its a crucial one. How should I tackle this ?

I would suggest pulling the `isDir` check out of the new validation function you added to `masonNew`, since that check is specific to `masonNew`.

ankingcodes commented:
@ben-albrecht  done that ! 
Okay, should some text be displayed after init succesfully creates all files ? If so, could u give some examples ?

ben-albrecht commented:
`mason new Foo` emits:

```
Created new library project: Foo
```

How about `mason init` does:

```
Initialized new library project: Foo
```

and ` mason init <path>` emits:

```
Initialized new library project in <path>: Foo
```

ankingcodes commented:
@ben-albrecht Thanks ! Added the comments.
Also added the tests for `mason init`. Please review once. 
It's failing one CI check though for chpl-docs. How does that work ? 
Could you point me to some documentation ?

ben-albrecht commented:
@ankingcodes - I see that Travis CI is complaining about using hard tabs instead of soft tabs (white space):

> Found tabs :-(

However, I am not sure if that's the cause of the actual CI failure. I suggest fixing that and seeing what Travis testing says.

ben-albrecht commented:
@Aniket21mathur -- I was able to get `testEigen.chpl` passing with the following modifications:

```diff
diff --git a/modules/packages/LinearAlgebra.chpl b/modules/packages/LinearAlgebra.chpl
index 7ab78c66c5..ee29ae0b78 100644
--- a/modules/packages/LinearAlgebra.chpl
+++ b/modules/packages/LinearAlgebra.chpl
@@ -1454,7 +1454,7 @@ proc eig(A: [] ?t, param left = false, param right = false)
     var vl, vr: [1..1, 1..n] t;
     if t == complex {
     LAPACK.geev(lapack_memory_order.row_major, 'N', 'N', copy, wr, vl, vr);
-    eigVals = w;
+    eigVals = wr;
     }else{
     LAPACK.geev(lapack_memory_order.row_major, 'N', 'N', copy, wr, wi, vl, vr);
     eigVals = convertToCplx(wr, wi);
@@ -1466,7 +1466,7 @@ proc eig(A: [] ?t, param left = false, param right = false)
     var vlcplx: if t == complex then vl else [1..n, 1..n] complex(numBits(t)*2);
     if t == complex {
     LAPACK.geev(lapack_memory_order.row_major, 'V', 'N', copy, wr, vl, vr);
-    eigVals = w;
+    eigVals = wr;
     vlcplx = vl;
     }else{
     LAPACK.geev(lapack_memory_order.row_major, 'V', 'N', copy, wr, wi, vl, vr);
@@ -1480,7 +1480,7 @@ proc eig(A: [] ?t, param left = false, param right = false)
     var vrcplx: if t == complex then vr else [1..n, 1..n] complex(numBits(t)*2);
     if t == complex {
     LAPACK.geev(lapack_memory_order.row_major, 'N', 'V', copy, wr, vl, vr);
-    eigVals = w;
+    eigVals = wr;
     vrcplx = vr;
     }else{
     LAPACK.geev(lapack_memory_order.row_major, 'N', 'V', copy, wr, wi, vl, vr);
@@ -1496,7 +1496,7 @@ proc eig(A: [] ?t, param left = false, param right = false)
     var vrcplx: if t == complex then vr else [1..n, 1..n] complex(numBits(t)*2);
     if t == complex {
     LAPACK.geev(lapack_memory_order.row_major, 'V', 'V', copy, wr, vl, vr);
-    eigVals = w;
+    eigVals = wr;
     vlcplx = vl;
     vrcplx = vr;
 ```

I did not see the internal error still, however.

ben-albrecht commented:
I checked your branch out and did some  testing. This is not at all comprehensive, but should give you an idea of the types of things you should check on your side as you develop. Ideally we would like to incorporate many of these checks into the automated testing:


```sh
mkdir empty
cd empty
mason init
# issue: chapel program created as: `src/..chpl`
```

```sh
cd empty
mason init empty
# Everything looks good!
```

```sh
mason new newPkg
mason init newPkg
# issue: We should probably emit an error if all files are present and 
#         mason init does not do anything
```

```sh
mason new newPkg
cd newPkg
rm -rf src
mason init
# issue: src/newPkg.chpl not created
```

```sh
mason new newPkg
cd newPkg
rm -rf .git
mason init
# issue: .git not created
```

```sh
mason new newPkg
cd newPkg
rm -rf Mason.toml
mason init
# Everything looks good!
```

```sh
mason new newPkg
cd newPkg
# remove 'name' field from Mason.toml 
mason init
# issue:
./MasonInit.chpl:48: error: attempt to dereference nil
```

ankingcodes commented:
@ben-albrecht Thanks for the review. In an attempt to recycle functions of `masonNew` for corner cases, I made silly mistakes on the base cases,

```
mason new newPkg
cd newPkg
remove 'name' field from Mason.toml 
mason init
issue:
./MasonInit.chpl:48: error: attempt to dereference nilbase cases. I've resolved them now. 
```
I have added error messages for this case. I hope that will do.

ben-albrecht commented on tools/mason/MasonInit.chpl:
```diff
@@ -0,0 +1,160 @@
+  }
+  if ToBeCreated.size == 1 {
+    throw new owned MasonError("Library project has already been initialised.");
+    exit(1);
```
This exit will not be reached if the error is thrown.

ben-albrecht commented on tools/mason/MasonInit.chpl:
```diff
@@ -0,0 +1,160 @@
+    exit(1);
+  }
+
+  for file in ToBeCreated{
```
I am not a huge fan of the `makeProjectFiles()` refactor, due to the reliance on special string values. What do you think about the following strategy?

Create a set of functions in `MasonNew` that create each component of the package:

```chpl
// Already exists
proc makeBasicToml(name, path=name)
proc gitInit(name) 
proc addGitIgnore(name)

// to add
proc makeSrcDir(name, path=name)
proc makeModule(name, path=Path.pathJoin(name, 'src'))
proc makeTestDir(name, path=name)
proc makeExampleDir(name, path=name)
```

`makeProjectFiles()` can call each of these individually, while this loop can do something like:

```chpl
const name = basename(path);

for file in ToBeCreated {
  select file {
    when '/Mason.toml' {
       makeBasicToml(name, path);
    }
    when '/.git' {
      gitInit(path, show=false);
    }
   // 
   // continue this pattern for each project file
   //
    otherwise {
      throw new owned MasonError('Unexpected file name to be generated.');
    }
  }
}
```

ankingcodes replied:
I did think about this too ! But I was trying my best not to modify previously written code in `MasonNew`. Thinking of this, I feel we should consider checking other `Mason` files for refactoring in future issues.

ben-albrecht commented on tools/mason/MasonInit.chpl:
```diff
@@ -0,0 +1,160 @@
+
+proc ValidateInit(path: string) throws {
+  var files = [ "/Mason.toml" , "/src" , "/test" , "/example", "/.git", ".gitignore" ];
+  var ToBeCreated : list(string);
```
should be camelCase:

```suggestion
  var toBeCreated : list(string);
```

ben-albrecht commented on test/mason/masonInit/masonInitPathTest.chpl:
```diff
@@ -0,0 +1,13 @@
+use FileSystem;
```
Could you add some more test cases, such as the ones I tried in the review? You could follow what I did like so:

```chpl
const newArgs = ['new', 'testSrc'];
masonNew(args)
runCommand(' rm -rf testSrc/src');
const initArgs = ['init', 'testSrc'];
masonInit(initArgs);
// check to make sure `src/` and `src/testSrc.chpl` were created
runCommand('rm -rf testSrc');
```

ankingcodes commented on tools/mason/MasonInit.chpl:
```diff
@@ -31,93 +31,130 @@ use FileSystem;
+      if file == "/.git" {
+        gitInit(path, show=false);
+      } 
+			else if file == ".gitignore" {
```
@ben-albrecht all hard tabs had been converted to spaces. It got passed by the Travis CI check.

ankingcodes commented on tools/mason/MasonInit.chpl:
```diff
@@ -0,0 +1,160 @@
+}
+
+proc ValidateInit(path: string) throws {
+  var files = [ "/Mason.toml" , "/src" , "/test" , "/example", "/.git", ".gitignore" ];
```
The prefixed slash is required to look for the files in a given path which may be `.` or `some/path`. So it's better we stick to this.

ankingcodes commented on tools/mason/MasonInit.chpl:
```diff
@@ -0,0 +1,160 @@
+  for idx in 1..files.size do {
+    var file = files(idx);
+    if file == "/Mason.toml" {
+      if isFile(path + file) {
```
It just seemed easier this way because we are only searching for a file from the files array in a directory. I would keep it this way if it's not absolutely necessary to use `Path.joinPath` here

ankingcodes commented:
@ben-albrecht made the changes you requested ! Also added the tests. Let me know if I need to add more.

ben-albrecht commented on tools/mason/MasonInit.chpl:
```diff
@@ -0,0 +1,174 @@
+    } 
+  }
+
+  for file in toBeCreated{
```
`file` is a type in the standard lib ([link](https://chapel-lang.org/docs/master/modules/standard/IO.html?highlight=file#IO.file)) - to avoid confusion, I'd suggest renaming, e.g. `masonFile`, `metafile`, `f`

ankingcodes commented on test/mason/masonInit/masonInitTest.chpl:
```diff
@@ -0,0 +1,20 @@
+  const pwd = getEnv('PWD');
+  if isDir(pwd + '/src'){
+      writeln("Init is correctly done");
+      runCommand('rm -rf src example test Mason.toml');
```
Added `.git` to list

ankingcodes commented:
@ben-albrecht sorry for delay. 
Modified the operations for the `Mason.toml` file and also added tests. 

> if no name found, error
> if no version found, error
> if no chplVersion found, add chplVersion = <current version>, otherwise leave it alone
> if no [dependencies] section found, add an empty one, otherwise leave it alone

also added the above conditions in `validateMasonFile()`
I agree that a `--show` tag would be helpful. Will be implementing that in a future PR. 
Other than that , I think this is ready for merge 👍

ankingcodes commented:
> While testing this out, I thought it'd be really nice to have a way to print exactly what mason init creates or modifies when running. I think we could add this via the --show flag (the verbose flag for mason commands). 

@ben-albrecht Why not use a `--interactive` tag as discussed in #14786 ? This would run a CLI questionnaire for mason, in which the inputs are given by the user for `name`,`version`,`author`,etc. are used up for creating the `Mason.toml` file. If user leaves fields vacant, mason uses default values for the fields.
Also, mason would show what it is creating or modifying.

ben-albrecht commented on tools/mason/MasonInit.chpl:
```diff
@@ -0,0 +1,199 @@
+      } 
+      else {
+        throw new owned MasonError("Directory does not exist:" + path +
+                              ". Did you mean 'mason new' to create a new project from scratch?"); 
```
Can we put `Did you mean 'mason new' to create a new project from scratch?`on a new line?

ben-albrecht commented:
> @ben-albrecht Why not use a `--interactive` tag as discussed in #14786 ? This would run a CLI questionnaire for mason, in which the inputs are given by the user for `name`,`version`,`author`,etc. are used up for creating the `Mason.toml` file. If user leaves fields vacant, mason uses default values for the fields.
Also, mason would show what it is creating or modifying. 

I think we could support both, and tackle them as separate efforts. `--show` is just the _verbose_ flag of mason, and it would make sense for all subcommands to support it to help debug issues / understand what mason is doing under the covers. `--interactive` would be great for `mason init` and `mason new`.

ankingcodes commented:
@ben-albrecht made the changes you requested. I'm working on the `--show` tag now. We can take care of `--interactive` in a separate PR. Also, I think we should also add `masonInitHelp` in `MasonHelp.chpl`, I missed it.

ankingcodes commented:
@ben-albrecht done with `--show` and `masonInitHelp`. Please review 👍

ankingcodes commented:
#14955 brought out another drawback of `mason init`. 
Apparently, we can make a project `chpl-project` using `mason init`.
Making a patch for this.

ben-albrecht commented on tools/mason/MasonHelp.chpl:
```diff
@@ -132,6 +132,16 @@ proc masonNewHelp() {
+proc masonInitHelp(){
+  writeln("Initializes a library project inside a given directory or path.");
+  writeln('Usage:');
+  writeln('    mason init [options] <directory name> || mason init [options]');
```
I would suggest putting these 2 on separate lines

ben-albrecht commented on tools/mason/MasonInit.chpl:
```diff
@@ -109,43 +116,49 @@ proc validateInit(path: string) throws {
     const name = basename(path);
     if metafile == "/.git" {
       gitInit(path, show=false);
```
Any subcommands that support `show` should get a `show=show` now.

ankingcodes commented:
Done with the requested changes and added documentation.

ben-albrecht commented on doc/rst/tools/mason/mason.rst:
```diff
@@ -41,6 +41,19 @@ To initialize a new mason package, run the ``mason new [ package name ] [ option
+``mason init [options]`` command outside or inside the project directory respectively. 
+
+For example, for an existing directory named MyPackage, 
+    
```
I think we will need a code-block for this to render, e.g.:

```
.. code-block: sh
    
    mason init MyPackage 

    # OR 

    cd MyPackage
    mason init

ankingcodes replied:
done 👍

ben-albrecht commented:
@ankingcodes - I noticed some errors in testing:

```
> start_test test/mason
...
> cat test/Logs/*.log.summary
[Test Summary - 200223.205226]
[Error matching program output for mason/masonInit/MasonInitTest5]
[Error matching program output for mason/masonInit/masonInitModifyToml]
[Error matching program output for mason/masonInit/masonInitPathTest]
[Error matching program output for mason/masonInit/masonInitTest]
[Error matching program output for mason/masonInit/masonInitTest2]
[Error matching program output for mason/masonInit/masonInitTest3]
[Error matching program output for mason/masonInit/masonInitTest4]
[Error cannot locate program output comparison file mason/masonInit/testSrc/src/testSrc.good]
[Summary: #Successes = 74 | #Failures = 8 | #Futures = 0 | #Warnings = 0 ]
[Summary: #Passing Suppressions = 0 | #Passing Futures = 0 ]
[END]
```

ankingcodes commented:
@ben-albrecht fixed the test errors. Please review it.

ankingcodes commented:
@ben-albrecht 
I deleted my local repo and cloned it again. After that I tried to build mason, but then I faced an error :
```
Building Mason... 
./MasonInit.chpl:219: In function 'addSection': 
./MasonInit.chpl:222: error: Cannot initialize associative array because element type unmanaged Toml is a non-nilable class
```
I hadn't encountered this before. Could you check for this & guide me ?

ben-albrecht commented:
@ankingcodes - There were changes merged into master that conflicted with this branch. It was not a trivial fix, so I went ahead and fixed it here: https://github.com/chapel-lang/chapel/pull/14985/files

ankingcodes commented:
@ben-albrecht thanks ! And I'll make sure to update my branch from next time.
